File: src/main/java/io/luna/game/model/mob/inter/OfferTradeInterface.java
package io.luna.game.model.mob.inter;

import com.google.common.collect.ImmutableList;
import io.luna.game.model.item.IndexedItem;
import io.luna.game.model.item.Inventory;
import io.luna.game.model.item.Item;
import io.luna.game.model.item.ItemContainer;
import io.luna.game.model.item.RefreshListener;
import io.luna.game.model.mob.Player;
import io.luna.game.model.mob.attr.AttributeValue;
import io.luna.net.msg.out.WidgetIndexedItemsMessageWriter;
import io.luna.net.msg.out.WidgetItemsMessageWriter;
import io.luna.util.LazyVal;

import java.util.List;

/**
 * An {@link InventoryOverlayInterface} implementation representing the offer trading screen.
 *
 * @author lare96 <http://github.com/lare96>
 */
public final class OfferTradeInterface extends InventoryOverlayInterface {

    /**
     * A {@link RefreshListener} that listens for items on the offer screen.
     */
    private final class OfferListener extends RefreshListener {

        /**
         * The player.
         */
        private final Player player;

        /**
         * Creates a new {@link Player}.
         *
         * @param player The player.
         */
        public OfferListener(Player player) {
            this.player = player;
        }

        @Override
        public void displayUpdate(ItemContainer items, List<IndexedItem> updateItems,
                                  WidgetIndexedItemsMessageWriter msg) {
            player.queue(msg); // Send to left panel.
            tradingWith.queue(new WidgetIndexedItemsMessageWriter(3416, updateItems)); // Send to right panel.
        }
    }

    /**
     * The player being traded with.
     */
    private final Player tradingWith;

    /**
     * The items being offered.
     */
    private final ItemContainer tradeItems = new ItemContainer(28, ItemContainer.StackPolicy.STANDARD, 3415);

    /**
     * The "trading_with" attribute.
     */
    private AttributeValue<Integer> tradingAttr;

    /**
     * The trading player's offer instance.
     */
    private LazyVal<OfferTradeInterface> otherOffer;

    /**
     * If the "Accept" button has been clicked.
     */
    private boolean accepted;

    /**
     * Creates a new {@link OfferTradeInterface}.
     *
     * @param tradingWith The player being traded with.
     */
    public OfferTradeInterface(Player tradingWith) {
        super(3323, 3321);
        this.tradingWith = tradingWith;
        otherOffer = new LazyVal<>(() -> tradingWith.getInterfaces().
                standardTo(OfferTradeInterface.class).
                orElseThrow(IllegalStateException::new));
    }

    @Override
    public void onOpen(Player player) {
        // Initialize variables.
        tradingAttr = player.getAttributes().get("trading_with");

        // Send text to trading interface.
        sendTradingWith(player);
        player.sendText("", 3431);
        player.sendText("Are you sure you want to make this trade?", 3535);

        // Refresh inventory to trade inventory.
        Inventory inventory = player.getInventory();
        inventory.setSecondaryRefresh(3322);
        inventory.refreshSecondary(player);

        // Clear left and right trade panels.
        WidgetItemsMessageWriter clearLeftMsg = new WidgetItemsMessageWriter(3415, ImmutableList.of());
        WidgetItemsMessageWriter clearRightMsg = new WidgetItemsMessageWriter(3416, ImmutableList.of());
        player.queue(clearLeftMsg);
        player.queue(clearRightMsg);

        // Set listeners.
        tradeItems.setListeners(new OfferListener(player));
    }

    @Override
    public void onClose(Player player) {
        // Trade was declined.
        tradingAttr.set(-1);
        player.getInventory().resetSecondaryRefresh();
        player.getInventory().addAll(tradeItems);

        tradingWith.getInterfaces().close();
    }

    @Override
    public void onReplace(Player player, AbstractInterface replace) {
        // If replacing interface isn't the confirm screen, decline.
        if (replace.getClass() != ConfirmTradeInterface.class) {
            onClose(player);
        }
    }

    /**
     * Adds an item to the offer screen.
     *
     * @param player The player.
     * @param index The inventory index of the item.
     * @param amount The amount to add.
     */
    public void add(Player player, int index, int amount) {
        Inventory inventory = player.getInventory();
        Item inventoryItem = inventory.get(index);

        // Item doesn't exist.
        if (inventoryItem == null) {
            return;
        }

        // Item can't be traded.
        if (!inventoryItem.getItemDef().isTradeable()) {
            player.sendMessage("This item cannot be traded.");
            return;
        }

        // Modify amount if there's less than requested.
        int hasAmount = inventoryItem.getItemDef().isStackable() ?
                inventoryItem.getAmount() :
                inventory.computeAmountForId(inventoryItem.getId());
        if (amount == -1 || amount > hasAmount) {
            amount = hasAmount;
        }

        // Add to offer screen.
        Item tradeItem = inventoryItem.withAmount(amount);
        if (inventory.remove(index, tradeItem)) {
            tradeItems.add(tradeItem);
            otherOffer.get().sendTradingWith(tradingWith);
            resetAccept(player);
        }
    }

    /**
     * Removes an item from the offer screen.
     *
     * @param player The player.
     * @param index The offer screen index of the item.
     * @param amount The amount to remove.
     */
    public void remove(Player player, int index, int amount) {
        Item tradeItem = tradeItems.get(index);

        // Item doesn't exist.
        if (tradeItem == null) {
            return;
        }

        // Modify amount if there's less than requested.
        int hasAmount = tradeItem.getItemDef().isStackable() ?
                tradeItem.getAmount() :
                tradeItems.computeAmountForId(tradeItem.getId());
        if (amount == -1 || amount > hasAmount) {
            amount = hasAmount;
        }

        // Remove from offer screen.
        Item removeItem = tradeItem.withAmount(amount);
        if (tradeItems.remove(removeItem)) {
            player.getInventory().add(removeItem);
            otherOffer.get().sendTradingWith(tradingWith);
            resetAccept(player);
        }
    }

    /**
     * Invoked when {@code player} clicks the "Accept" button on the offer screen.
     *
     * @param player The player.
     */
    public void accept(Player player) {
        if (accepted) {
            return;
        }

        // Trading player doesn't have enough inventory space.
        if (tradingWith.getInventory().computeRemainingSize() < tradeItems.size()) {
            player.sendMessage(tradingWith.getUsername() + " does not have enough inventory space for your items.");
            return;
        }

        if (otherOffer.get().accepted) {
            // Both players have accepted, open confirmation screen.
            ConfirmTradeInterface confirm = new ConfirmTradeInterface(this);
            ConfirmTradeInterface otherConfirm = new ConfirmTradeInterface(otherOffer.get());
            player.getInterfaces().open(confirm);
            tradingWith.getInterfaces().open(otherConfirm);
        } else {
            // Accept trade, wait for other player to accept.
            accepted = true;
            player.sendText("Waiting for other player...", 3431);
            tradingWith.sendText("Other player has accepted", 3431);
        }
    }

    /**
     * Displays the trading Player's name and remaining free space.
     *
     * @param player The player.
     */
    private void sendTradingWith(Player player) {
        String username = tradingWith.getUsername();
        StringBuilder sb = new StringBuilder(username.length() + 5);

        // Append username and crown.
        sb.append(username);
        switch (tradingWith.getRights()) {
            case MODERATOR:
                sb.append("@cr1@");
            case DEVELOPER:
            case ADMINISTRATOR:
                sb.append("@cr2@");
        }

        // Send widget text to player.
        int remaining = tradingWith.getInventory().computeRemainingSize();
        player.sendText("Trading with: " + sb + " who has @gre@" + remaining + " free slots", 3417);
    }

    /**
     * Resets the accept button click for {@code player} and the trading Player.
     *
     * @param player The player.
     */
    private void resetAccept(Player player) {
        accepted = false;
        otherOffer.get().accepted = false;
        player.sendText("", 3431);
        tradingWith.sendText("", 3431);
    }

    /**
     * @return The items being offered.
     */
    ItemContainer getTradeItems() {
        return tradeItems;
    }

    /**
     * @return The player being traded with.
     */
    Player getTradingWith() {
        return tradingWith;
    }
}

File: plugins/world/player/item/trading/tradeScreen/OfferTradeInterface.kt
package world.player.item.trading.tradeScreen

import api.predef.*
import io.luna.game.model.item.IndexedItem
import io.luna.game.model.item.Item
import io.luna.game.model.item.ItemContainer
import io.luna.game.model.item.RefreshListener
import io.luna.game.model.mob.Player
import io.luna.game.model.mob.inter.AbstractInterface
import io.luna.game.model.mob.inter.InventoryOverlayInterface
import io.luna.net.msg.out.WidgetIndexedItemsMessageWriter
import io.luna.net.msg.out.WidgetItemsMessageWriter

/**
 * An [InventoryOverlayInterface] implementation representing the offer trading screen.
 */
class OfferTradeInterface(val other: Player) : InventoryOverlayInterface(3323, 3321) {

    /**
     * A [RefreshListener] that listens for items on the offer screen.
     */
    inner class OfferListener(private val plr: Player) : RefreshListener() {

        override fun displayUpdate(items: ItemContainer, updateItems: List<IndexedItem>,
                                   msg: WidgetIndexedItemsMessageWriter) {

            plr.queue(msg) // Send to left panel.
            other.queue(WidgetIndexedItemsMessageWriter(3416, updateItems)) // Send to right panel.
        }
    }

    /**
     * The items being offered.
     */
    val items = ItemContainer(28, ItemContainer.StackPolicy.STANDARD, 3415)

    /**
     * The trading player's offer instance (load lazily so its not initialized in the constructor).
     */
    val otherOffer by lazyVal {
        other.interfaces.get(OfferTradeInterface::class)!!
    }

    /**
     * If the "Accept" button has been clicked.
     */
    var accepted = false

    override fun onOpen(plr: Player) {
        sendTradingWith(plr)
        plr.sendText("", 3431)
        plr.sendText("Are you sure you want to make this trade?", 3535)

        // Refresh inventory to trade inventory.
        plr.inventory.setSecondaryRefresh(3322)
        plr.inventory.refreshSecondary(plr)

        // Clear left and right trade panels.
        val clearLeftMsg = WidgetItemsMessageWriter(3415, listOf())
        val clearRightMsg = WidgetItemsMessageWriter(3416, listOf())
        plr.queue(clearLeftMsg)
        plr.queue(clearRightMsg)

        items.setListeners(OfferListener(plr))
    }

    override fun onClose(plr: Player) {
        // Trade was declined.
        plr.tradingWith = -1

        plr.inventory.resetSecondaryRefresh()
        plr.inventory.addAll(items)

        other.interfaces.close()
    }

    override fun onReplace(plr: Player, replace: AbstractInterface) {
        // If replacing interface isn't the confirm screen, decline.
        if (replace::class != ConfirmTradeInterface::class) {
            onClose(plr)
        }
    }

    /**
     * Adds [amount] item from inventory slot [index] to the offer screen.
     */
    fun add(plr: Player, index: Int, amount: Int) {
        val inv = plr.inventory
        val item = inv.get(index) ?: return

        if (!itemDef(item.id).isTradeable) {
            plr.sendMessage("This item cannot be traded.")
            return
        }

        val addItem = reduceAmount(item, amount, inv)
        if (inv.remove(index, addItem)) {
            items.add(addItem)
            otherOffer.sendTradingWith(other)
            resetAccept(plr)
        }
    }

    /**
     * Removes [amount] item from trade screen slot [index] to the inventory.
     */
    fun remove(plr: Player, index: Int, amount: Int) {
        val item = items.get(index) ?: return

        val removeItem = reduceAmount(item, amount, items)
        if (items.remove(index, removeItem)) {
            plr.inventory.add(removeItem)
            otherOffer.sendTradingWith(other)
            resetAccept(plr)
        }
    }

    /**
     * Invoked when [plr] clicks the "Accept" button on the offer screen.
     */
    fun accept(plr: Player) {
        if (accepted) {
            return
        }

        if (other.inventory.computeRemainingSize() < items.size()) {
            plr.sendMessage("${other.username} does not have enough inventory space for your items.")
            return
        }

        if (otherOffer.accepted) {
            val confirm = ConfirmTradeInterface(this)
            val otherConfirm = ConfirmTradeInterface(otherOffer)
            plr.interfaces.open(confirm)
            other.interfaces.open(otherConfirm)
        } else {
            accepted = true
            plr.sendText("Waiting for other player...", 3431)
            other.sendText("Other player has accepted", 3431)
        }
    }

    /**
     * Returns an new amount instance of [item] with [reqAmt] reduced to the amount currently contained
     * within [container].
     */
    private fun reduceAmount(item: Item, reqAmt: Int, container: ItemContainer): Item {
        val id = item.id
        val amount = item.amount
        val hasAmount = if (itemDef(id).isStackable) amount else container.computeAmountForId(id)
        return when {
            // Modify amount if there's less in than requested.
            reqAmt == -1 || reqAmt > hasAmount -> item.withAmount(hasAmount)
            else -> item.withAmount(reqAmt)
        }
    }

    /**
     * Displays the trading Player's name and remaining free space for [plr].
     */
    private fun sendTradingWith(plr: Player) {
        val username = other.username
        val sb = StringBuilder(username.length + 5)
        sb.append(username)
        when (other.rights) {
            RIGHTS_MOD -> sb.append("@cr1@")
            RIGHTS_ADMIN, RIGHTS_DEV -> sb.append("@cr2@")
            else -> Unit
        }

        val remaining = other.inventory.computeRemainingSize()
        plr.sendText("Trading with: $sb who has @gre@$remaining free slots", 3417)
    }

    /**
     * Resets the accept button click for [plr] and the trading Player.
     */
    private fun resetAccept(plr: Player) {
        accepted = false
        otherOffer.accepted = false
        plr.sendText("", 3431)
        other.sendText("", 3431)
    }
}